src/index.ts

```diff
@@ -203,158 +232,217 @@ function getPropertyNames(schema: Schema) {
+    } else if (helpers.isArray(prop)) {
+      // unofficial extension - support fixed width types
+      const arrayType = prop.xtpType as ArrayType
+      switch (arrayType.elementType.kind) {
+        case 'byte':
+        case 'int32':
+        case 'int64':
+        case 'float':
+        case 'double':
+          return v2ToCppTypeXInner(arrayType.elementType, '')
       }
     }
+  } else if (prop.contentType === 'text/plain; charset=utf-8') {
+    if (helpers.isString(prop) || helpers.isBuffer(prop)) {
+      return 'char'
+    }
   }
-  throw new Error("not sure what the handle type should be for " + prop.type + ' encoded as ' + prop.contentType);
+  throw new Error("not sure what the handle type should be for " + prop.xtpType.kind + ' encoded as ' + prop.contentType);
 }
 
 function getHandleAccessor(prop: Parameter) {
   if (getHandleType(prop) != 'char') {
     return 'vec';
-  } else if (!prop['$ref'] && prop.contentType === 'text/plain; charset=utf-8' && prop.type === 'buffer') {
+  } else if (prop.contentType === 'text/plain; charset=utf-8' && helpers.isBuffer(prop)) {
     return 'vec';
   }
   return 'string';
 }
 
 function getJSONDecodeType(param: Parameter) {
-  return toStorageType(param, '');
+  return v2ToStorageType(param.xtpType, '')
 }
 
 function derefIfNotOptionalPointer(param: Parameter) {
-  if (param.nullable || !isLargeType(param)) {
+  if (param.xtpType.nullable || !v2IsLargeType(param.xtpType)) {
     return '';
   }
   return '*';
 }
 
 function needsNullCheck(param: Parameter) {
-  return !param.nullable && isLargeType(param);
+  return !param.xtpType.nullable && v2IsLargeType(param.xtpType)
 }
 
-function hasUntypedObject(objects: any) {
-  for (const object of objects) {
-    for (const prop of object.properties) {
-      if (!prop.$ref && prop.type === 'object') {
-        return true;
-      }
+function objectHasBuffer(schema: Schema) {
+  const object = schema.xtpType as ObjectType
+  for (const prop of object.properties) {
+    if (prop.kind === 'buffer') {
+      return true
     }
   }
-  return false;
+  return false
+}
+
+function isPropertyRequired(property: Property) {
+  return property.required === undefined || property.required
 }
 
-function objectHasBuffer(object: Schema) {
+function numRequiredProperties(object: Schema) {
+  let count = 0;
   for (const prop of object.properties) {
-    if (!prop.$ref && prop.type === 'buffer') {
-      return true;
+    if (isPropertyRequired(prop)) {
+      count++
     }
   }
+  return count
 }
 
-function usesBuffer(objects: any) {
-  for (const object of objects) {
-    if (objectHasBuffer(object)) {
-      return true;
+function sortedProperties(object: Schema) {
+  const properties: Property[] = []
+  let curRequiredProperties = 0
+  for (const prop of object.properties) {
+    if (!isPropertyRequired(prop)) {
+      properties.push(prop)
+      continue;
     }
+    properties.splice(curRequiredProperties, 0, prop)
+    curRequiredProperties++
```

Comment:
This is just golf, but it might play nicer with the JS allocator:

```js
const properties: Property[] = []
const requiredProperties: Property[] = []
for (const prop of object.properties) {
  const target = (isPropertyRequired(prop) ? requiredProperties : properties)
  target.push(prop) 
}
return [...requiredProperties, ...properties]
```

Reply:
Applied in e89d637, thanks I like that it's easier to follow too!